Fix plugin issues when disabling/reenabling plugins

StraToN opened the pull request:
Fixes https://github.com/godot-escoria/escoria-issues/issues/216
"escoria" singleton/autoload now only holds references to managers and only instances "atomic" managers.
Escoria scene is now instanced by main_scene. 
New logger added.

BHSDuncan commented on addons/escoria-core/game/esc_logger.gd:
```diff
@@ -0,0 +1,193 @@
+
+	# Constructor
+	func _init():
+		_log_level = _level_map[ESCProjectSettingsManager.get_setting(
+			ESCProjectSettingsManager.LOG_LEVEL.to_upper()
+		)]
```
```suggestion
		_log_level = _level_map[ESCProjectSettingsManager.get_setting(
			ESCProjectSettingsManager.LOG_LEVEL
		).to_upper()]
```

I think the idea here is to uppercase the value that the project setting returns for use in the log level dictionary.

balloonpopper commented on addons/escoria-core/game/core-scripts/esc/commands/camera_set_zoom_height.gd:
```diff
@@ -30,11 +30,10 @@ func validate(arguments: Array):
-			"camera_set_zoom_height: invalid height",
-			[
-				"Can't zoom to a negative height %d" % arguments[0]
-			]
+		escoria.logger.error(
+			self,
+			get_command_name() + ": invalid height" +
+					"Can't zoom to a negative height %d" % arguments[0]
 		)
```
Is a height of 0 valid? Should this check be <= 0?

balloonpopper commented on addons/escoria-core/game/core-scripts/esc/commands/dec_global.gd:
```diff
@@ -27,11 +27,10 @@ func validate(arguments: Array):
-		escoria.logger.report_errors(
-			"dec_global: invalid global",
-			[
+		escoria.logger.error(
+			self,
+			get_command_name() + ": invalid global." +
 				"Global %s didn't have an integer value." % arguments[0]
```
```suggestion
		escoria.logger.error(
			self,
			get_command_name() + ": invalid global." +
				"Global %s isn't an integer value." % arguments[0]
```

balloonpopper commented on addons/escoria-core/game/core-scripts/esc/commands/enable_terrain.gd:
```diff
@@ -36,11 +36,10 @@ func run(command_params: Array) -> int:
-		escoria.logger.report_errors(
-			"EnableTerrainCommand.run: Can not find terrain node",
-			[
+		escoria.logger.error(
+			self,
+			get_command_name() + ": Can not find terrain node" +
 				"Terrain node %s could not be found" % name
```
```suggestion
		escoria.logger.error(
			self,
			get_command_name() + ": Can not find terrain node" +
				"Terrain node %s could not be found." % name
```

balloonpopper commented on addons/escoria-core/game/core-scripts/behaviors/esc_movable.gd:
```diff
@@ -158,15 +158,10 @@ func _perform_walk_orientation(angle: float):
-				"proceed movement."
-			],
-			true
+		escoria.logger.warn(
+			self,
+			"Character %s has no animation %s\nBypassing missing animation and proceed movement."
+					% [parent.global_id, animation_to_play]
```
proceed movement doesn't make sense in this error message. Should it be
```suggestion
		escoria.logger.warn(
			self,
			"Character %s has no animation %s\nBypassing the missing animation and movement command."
					% [parent.global_id, animation_to_play]
```

balloonpopper commented on addons/escoria-core/game/core-scripts/behaviors/esc_movable.gd:
```diff
@@ -181,21 +176,21 @@ func teleport(target: Node) -> void:
-				parent.global_position
-			]
+			self,
+			"Object %s is teleported at position %s" 
```
```suggestion
			"Object %s is teleported to position %s." 
```

balloonpopper commented on addons/escoria-core/game/core-scripts/behaviors/esc_movable.gd:
```diff
@@ -181,21 +176,21 @@ func teleport(target: Node) -> void:
-			"Object %s teleported at position %s" %
-			[parent.global_id, str(target.global_position)]
+			self,
+			"Object %s teleported at position %s"
```
```suggestion
			"Object %s teleported to position %s."
```

balloonpopper commented on addons/escoria-core/game/core-scripts/behaviors/esc_movable.gd:
```diff
@@ -181,21 +176,21 @@ func teleport(target: Node) -> void:
-		escoria.logger.report_errors(
-			"ESCMovable#teleport()",
-			["Couldn't understand how to manage teleport Target %s" % target]
+		escoria.logger.error(
+			self,
+			"Couldn't understand how to manage teleport Target %s" % target
```
Please confirm my suggested error matches what the code is trying to do.
```suggestion
		escoria.logger.error(
			self,
			"Target %s could not be teleported. Please configure the interact position parameter or create a child ESCLocation node." % target
```

balloonpopper commented on addons/escoria-core/game/core-scripts/behaviors/esc_movable.gd:
```diff
@@ -206,8 +201,9 @@ func teleport(target: Node) -> void:
 	escoria.logger.info(
-		"Object %s teleported to position %s" %
-		[parent.global_id, str(target)]
+		self,
+		"Object %s teleported to position %s" 
+				% [parent.global_id, str(target)]
```
```suggestion
		self,
		"Object %s teleported to position %s." 
				% [parent.global_id, str(target)]
```

balloonpopper commented on addons/escoria-core/game/core-scripts/behaviors/esc_movable.gd:
```diff
@@ -287,13 +283,15 @@ func walk_stop(pos: Vector2) -> void:
 
 	if walk_context.target_object:
 		escoria.logger.debug(
+			self,
 			"%s arrived at %s" % [
 				parent.global_id,
 				walk_context.target_object.global_id
 			]
 		)
 	else:
 		escoria.logger.debug(
+			self,
 			"%s arrived at %s" % [
 				parent.global_id,
 				walk_context.target_position
```
```suggestion
	if walk_context.target_object:
		escoria.logger.debug(
			self,
			"%s arrived at %s." % [
				parent.global_id,
				walk_context.target_object.global_id
			]
		)
	else:
		escoria.logger.debug(
			self,
			"%s arrived at %s." % [
				parent.global_id,
				walk_context.target_position
```

balloonpopper commented on addons/escoria-core/game/core-scripts/behaviors/esc_movable.gd:
```diff
@@ -376,9 +375,9 @@ func _get_dir_deg(deg: int, animations: ESCAnimationResource) -> int:
-		escoria.logger.report_errors(
-			"esc_movable.gd:_get_dir_deg()",
-			["No direction found for " + str(deg)]
+		escoria.logger.error(
+			self,
+			"No direction found for %s" + str(deg)
```
```suggestion
		escoria.logger.error(
			self,
			"No animation has been configured for angle %s." + str(deg)
```

balloonpopper commented on addons/escoria-core/game/core-scripts/behaviors/esc_movable.gd:
```diff
@@ -414,9 +413,9 @@ func _is_angle_in_interval(
-		escoria.logger.report_errors(
-			"movable.gd:set_angle()",
-			["Invalid degree to turn to " + str(deg)]
+		escoria.logger.error(
+			self,
+			"Invalid degree to turn to %s" % str(deg)
```
```suggestion
		escoria.logger.error(
			self,
			"Invalid degree to turn to : %s. Valid angles are between 0 and 360." % str(deg)
```

balloonpopper commented on addons/escoria-core/game/core-scripts/esc/commands/queue_event.gd:
```diff
@@ -31,39 +31,31 @@ func validate(arguments: Array):
-			[
-				"Object with global id %s not found" % arguments[0]
-			]
+		escoria.logger.error(
+			self,
+			"Object with global id %s not found" % arguments[0]
```
```suggestion
		escoria.logger.error(
			self,
			"Object with global id %s not found." % arguments[0]
```

balloonpopper commented on addons/escoria-core/game/core-scripts/esc/commands/queue_event.gd:
```diff
@@ -31,39 +31,31 @@ func validate(arguments: Array):
-			[
-				"Object with global id %s has no ESC script" % arguments[0]
-			]
+		escoria.logger.error(
+			self,
+			"Object with global id %s has no ESC script" % arguments[0]
```
```suggestion
		escoria.logger.error(
			self,
			"Object with global id %s has no ESC script." % arguments[0]
```

balloonpopper commented on addons/escoria-core/game/core-scripts/esc/commands/queue_event.gd:
```diff
@@ -31,39 +31,31 @@ func validate(arguments: Array):
-			[
-				"Event with name %s not found" % arguments[1]
-			]
+		escoria.logger.error(
+			self,
+			"Event with name %s not found" % arguments[1]
```
```suggestion
		escoria.logger.error(
			self,
			"Event with name %s not found." % arguments[1]
```

balloonpopper commented on addons/escoria-core/game/core-scripts/esc/commands/queue_resource.gd:
```diff
@@ -28,9 +28,10 @@ func validate(arguments: Array) -> bool:
-		escoria.logger.report_errors(
-			"queue_resource: Invalid resource",
-			["Resource %s was not found" % arguments[0]]
+		escoria.logger.error(
+			self,
+			get_command_name() + ": Invalid resource. " +
+				"Resource %s was not found" % arguments[0]
```
```suggestion
		escoria.logger.error(
			self,
			get_command_name() + ": Invalid resource. " +
				"Resource %s was not found." % arguments[0]
```

balloonpopper commented on addons/escoria-core/game/core-scripts/esc/commands/say.gd:
```diff
@@ -115,9 +110,7 @@ func run(command_params: Array) -> int:
-		[
-			"Interrupt() function not implemented"
-		]
+	escoria.logger.warn(
+		self,
+		"[%s] Interrupt() function not implemented" % get_command_name()
```
```suggestion
	escoria.logger.warn(
		self,
		"[%s] Interrupt() function not implemented." % get_command_name()
```

balloonpopper commented on addons/escoria-core/game/core-scripts/esc/commands/sched_event.gd:
```diff
@@ -32,23 +32,21 @@ func validate(arguments: Array):
-			[
-				"Object with global id %s not found" % arguments[1]
-			]
+		escoria.logger.error(
+			self,
+			get_command_name() + ": invalid object. " + 
+					"Object with global id %s not found" % arguments[1]
```
```suggestion
		escoria.logger.error(
			self,
			get_command_name() + ": invalid object. " + 
					"Object with global id %s not found." % arguments[1]
```

balloonpopper commented on addons/escoria-core/game/core-scripts/esc/commands/sched_event.gd:
```diff
@@ -32,23 +32,21 @@ func validate(arguments: Array):
-		escoria.logger.report_errors(
-			"sched_event: invalid object event",
-			[
+		escoria.logger.error(
+			self,
+			get_command_name() + ": invalid object event." +
 				"Object with global id %s has no event %s" % [
```
```suggestion
		escoria.logger.error(
			self,
			get_command_name() + ": invalid object event." +
				"Object with global id %s has no event %s." % [
```

balloonpopper commented on addons/escoria-core/game/core-scripts/esc/commands/set_active.gd:
```diff
@@ -28,11 +28,10 @@ func validate(arguments: Array):
-			[
-				"Object with global id %s not found" % arguments[0]
-			]
+		escoria.logger.error(
+			self,
+			get_command_name() + ": invalid object" +
+					"Object with global id %s not found" % arguments[0]
```
```suggestion
		escoria.logger.error(
			self,
			get_command_name() + ": invalid object" +
					"Object with global id %s not found." % arguments[0]
```

balloonpopper commented on addons/escoria-core/game/core-scripts/esc/commands/set_angle.gd:
```diff
@@ -33,11 +33,10 @@ func validate(arguments: Array):
-			[
-				"Object with global id %s not found" % arguments[0]
-			]
+		escoria.logger.error(
+			self,
+			get_command_name() + ": invalid object. " + 
+			"Object with global id %s not found" % arguments[0]
```
```suggestion
		escoria.logger.error(
			self,
			get_command_name() + ": invalid object. " + 
			"Object with global id %s not found." % arguments[0]
```

balloonpopper commented on addons/escoria-core/game/core-scripts/esc/commands/set_animations.gd:
```diff
@@ -27,19 +27,17 @@ func validate(arguments: Array):
-		escoria.logger.report_errors(
-			"set_animations: invalid object",
-			[
+		escoria.logger.error(
+			self,
+			get_command_name() + ": invalid object" +
 				"Object with global id %s not found" % arguments[0]
```
```suggestion
		escoria.logger.error(
			self,
			get_command_name() + ": invalid object" +
				"Object with global id %s not found." % arguments[0]
```

balloonpopper commented on addons/escoria-core/game/core-scripts/esc/commands/set_animations.gd:
```diff
@@ -27,19 +27,17 @@ func validate(arguments: Array):
-		escoria.logger.report_errors(
-			"set_animations: invalid animations",
-			[
+		escoria.logger.error(
+			self,
+			get_command_name() + ": invalid animations" +
 				"The animation resource %s was not found" % arguments[1]
```
```suggestion
		escoria.logger.error(
			self,
			get_command_name() + ": invalid animations" +
				"The animation resource %s was not found." % arguments[1]
```

balloonpopper commented on addons/escoria-core/game/core-scripts/esc/commands/set_interactive.gd:
```diff
@@ -27,11 +27,10 @@ func validate(arguments: Array):
-		escoria.logger.report_errors(
-			"set_interactive: invalid object",
-			[
+		escoria.logger.error(
+			self,
+			get_command_name() + ": invalid object. ",
 				"Object with global id %s not found" % arguments[0]
```
```suggestion
		escoria.logger.error(
			self,
			get_command_name() + ": invalid object. ",
				"Object with global id %s not found." % arguments[0]
```

balloonpopper commented on addons/escoria-core/game/core-scripts/esc/commands/set_speed.gd:
```diff
@@ -26,11 +26,10 @@ func validate(arguments: Array):
-		escoria.logger.report_errors(
-			"set_speed: invalid object",
-			[
+		escoria.logger.error(
+			self,
+			get_command_name() + ": invalid object. " +
 				"Object with global id %s not found" % arguments[0]
```
```suggestion
		escoria.logger.error(
			self,
			get_command_name() + ": invalid object. " +
				"Object with global id %s not found." % arguments[0]
```

balloonpopper commented on addons/escoria-core/game/core-scripts/esc/commands/show_menu.gd:
```diff
@@ -31,11 +31,9 @@ func validate(arguments: Array):
-			[
-				"menu %s is invalid" % arguments[0]
-			]
+		escoria.logger.error(
+			self,
+			get_command_name() + ": menu %s is invalid" % arguments[0]
```
```suggestion
		escoria.logger.error(
			self,
			get_command_name() + ": menu %s is invalid." % arguments[0]
```

balloonpopper commented on addons/escoria-core/game/core-scripts/esc/commands/slide.gd:
```diff
@@ -35,19 +35,17 @@ func validate(arguments: Array):
-			[
-				"Object with global id %s not found" % arguments[0]
-			]
+		escoria.logger.error(
+			self,
+			get_command_name() + ": invalid first object. " +
+			"Object with global id %s not found" % arguments[0]
```
```suggestion
		escoria.logger.error(
			self,
			get_command_name() + ": invalid first object. " +
			"Object with global id %s not found." % arguments[0]
```

balloonpopper commented on addons/escoria-core/game/core-scripts/esc/commands/slide.gd:
```diff
@@ -35,19 +35,17 @@ func validate(arguments: Array):
-			[
-				"Object with global id %s not found" % arguments[1]
-			]
+		escoria.logger.error(
+			self,
+			get_command_name() + ": invalid second object. " +
+			"Object with global id %s not found" % arguments[1]
```
```suggestion
		escoria.logger.error(
			self,
			get_command_name() + ": invalid second object. " +
			"Object with global id %s not found." % arguments[1]
```

balloonpopper commented on addons/escoria-core/game/core-scripts/esc/commands/spawn.gd:
```diff
@@ -31,27 +31,21 @@ func validate(arguments: Array):
-			[
-				"global_id %s is invalid" % arguments[0]
-			]
+		escoria.logger.error(
+			self,
+			get_command_name() + " global_id %s is invalid" % arguments[0]
```
```suggestion
		escoria.logger.error(
			self,
			get_command_name() + " global_id (%s) is invalid. The global_id was either empty or is reserved." % arguments[0]
```

balloonpopper commented on addons/escoria-core/game/core-scripts/esc/commands/spawn.gd:
```diff
@@ -31,27 +31,21 @@ func validate(arguments: Array):
-			[
-				"Scene with path %s not found" % arguments[1]
-			]
+		escoria.logger.error(
+			self,
+			get_command_name() + " Invalid scene path: %s not found" % arguments[1]
```
```suggestion
		escoria.logger.error(
			self,
			get_command_name() + " Invalid scene path: %s not found." % arguments[1]
```

balloonpopper commented on addons/escoria-core/game/core-scripts/esc/commands/spawn.gd:
```diff
@@ -31,27 +31,21 @@ func validate(arguments: Array):
-			[
-				"Object with global id %s not found" % arguments[3]
-			]
+		escoria.logger.error(
+			self,
+			get_command_name() + " invalid object: Object with global id %s not found" % arguments[3]
```
```suggestion
		escoria.logger.error(
			self,
			get_command_name() + " invalid object: Object with global id %s not found." % arguments[3]
```

balloonpopper commented on addons/escoria-core/game/core-scripts/esc/commands/spawn.gd:
```diff
@@ -83,11 +77,10 @@ func run(command_params: Array) -> int:
-			[
-				"Failed loading scene %s" % command_params[1]
-			]
+		escoria.logger.error(
+			self,
+			"[%s]: Invalid scene. Failed loading scene %s"
+					% [get_command_name(), command_params[1]]
```
```suggestion
		escoria.logger.error(
			self,
			"[%s]: Invalid scene. Failed to load scene %s."
					% [get_command_name(), command_params[1]]
```

balloonpopper commented on addons/escoria-core/game/core-scripts/esc/commands/stop_snd.gd:
```diff
@@ -35,9 +35,10 @@ func validate(arguments: Array):
-		escoria.logger.report_errors(
-			"stop_snd: invalid sound player",
-			["Sound player %s not registered" % arguments[0]]
+		escoria.logger.error(
+			self,
+			get_command_name() + ": invalid sound player" +
+					"Sound player %s not registered" % arguments[0]
```
```suggestion
		escoria.logger.error(
			self,
			get_command_name() + ": invalid sound player" +
					"Sound player %s not registered." % arguments[0]
```

balloonpopper commented on addons/escoria-core/game/core-scripts/esc/commands/teleport.gd:
```diff
@@ -28,19 +28,17 @@ func validate(arguments: Array):
-		escoria.logger.report_errors(
-			"teleport: invalid first object",
-			[
+		escoria.logger.error(
+			self,
+			get_command_name() + ": invalid first object. " +
 				"Object with global id %s not found" % arguments[0]
```
```suggestion
		escoria.logger.error(
			self,
			get_command_name() + ": invalid first object. " +
				"Object with global id %s not found" % arguments[0]
```
```suggestion
		escoria.logger.error(
			self,
			get_command_name() + ": invalid first object. " +
				"Object to teleport with global id %s not found." % arguments[0]
```

balloonpopper commented on addons/escoria-core/game/core-scripts/esc/commands/teleport.gd:
```diff
@@ -28,19 +28,17 @@ func validate(arguments: Array):
-		escoria.logger.report_errors(
-			"teleport: invalid second object",
-			[
+		escoria.logger.error(
+			self,
+			get_command_name() + ": invalid second object. " +
 				"Object with global id %s not found" % arguments[1]
```
```suggestion
		escoria.logger.error(
			self,
			get_command_name() + ": invalid second object. " +
				"Destination location to teleport to with global id %s not found." % arguments[1]
```

balloonpopper commented on addons/escoria-core/game/core-scripts/esc/commands/teleport_pos.gd:
```diff
@@ -28,11 +28,10 @@ func validate(arguments: Array):
-		escoria.logger.report_errors(
-			"teleport_pos: invalid first object",
-			[
+		escoria.logger.error(
+			self,
+			get_command_name() + ": invalid first object. " + 
 				"Object with global id %s not found" % arguments[0]
```
```suggestion
		escoria.logger.error(
			self,
			get_command_name() + ": invalid first object. " + 
				"Object to teleport with global id %s not found." % arguments[0]
```

balloonpopper commented on addons/escoria-core/game/core-scripts/esc/commands/transition.gd:
```diff
@@ -30,19 +30,17 @@ func validate(arguments: Array):
-			[
-				"transition with name '%s' doesn't exist" % arguments[0]
-			]
+		escoria.logger.error(
+			self,
+			get_command_name() + ": argument invalid. Transition with name '%s' doesn't exist" 
+					% arguments[0]
```
```suggestion
		escoria.logger.error(
			self,
			get_command_name() + ": argument invalid. Transition with name '%s' doesn't exist." 
					% arguments[0]
```

balloonpopper commented on addons/escoria-core/game/core-scripts/esc/commands/transition.gd:
```diff
@@ -30,19 +30,17 @@ func validate(arguments: Array):
-		escoria.logger.report_errors(
-			"transition: argument invalid",
-			[
+		escoria.logger.error(
+			self,
+			get_command_name() + ": argument invalid" +
 				"'in' or 'out' expected, but got '%s'" % arguments[1]
```
```suggestion
		escoria.logger.error(
			self,
			get_command_name() + ": argument invalid" +
				"Transition type 'in' or 'out' expected, but '%s' was provided." % arguments[1]
```

balloonpopper commented on addons/escoria-core/game/core-scripts/esc/commands/transition.gd:
```diff
@@ -58,19 +56,27 @@ func run(command_params: Array) -> int:
 
-	escoria.logger.debug("Starting transition #%s [%s, %s]"
-		% [transition_id, command_params[0], command_params[1]])
+	escoria.logger.debug(
+		self,
+		"Starting transition #%s [%s, %s]"
+				% [transition_id, command_params[0], command_params[1]]
+	)
```
```suggestion
	escoria.logger.debug(
		self,
		"Starting transition #%s [%s, %s]."
				% [transition_id, command_params[0], command_params[1]]
	)
```

balloonpopper commented on addons/escoria-core/game/core-scripts/esc/commands/transition.gd:
```diff
@@ -58,19 +56,27 @@ func run(command_params: Array) -> int:
 		pass
-	escoria.logger.debug("Ending transition #%s [%s, %s]"
-		% [transition_id, command_params[0], command_params[1]])
+	escoria.logger.debug(
+		self,
+		"Ending transition #%s [%s, %s]"
+				% [transition_id, command_params[0], command_params[1]])
```
```suggestion
	escoria.logger.debug(
		self,
		"Ending transition #%s [%s, %s]."
				% [transition_id, command_params[0], command_params[1]])
```

balloonpopper commented on addons/escoria-core/game/core-scripts/esc/commands/turn_to.gd:
```diff
@@ -35,19 +35,16 @@ func validate(arguments: Array):
-			[
-				"Object with global id %s not found" % arguments[0]
-			]
+		escoria.logger.error(
+			self,
+			get_command_name() + ": Invalid object: object with global id %s not found" % arguments[0]
```
```suggestion
		escoria.logger.error(
			self,
			get_command_name() + ": The object to turn with global id %s was not found." % arguments[0]
```

balloonpopper commented on addons/escoria-core/game/core-scripts/esc/commands/turn_to.gd:
```diff
@@ -35,19 +35,16 @@ func validate(arguments: Array):
-			[
-				"Object with global id %s not found" % arguments[1]
-			]
+		escoria.logger.error(
+			self,
+			get_command_name() + ": Invalid target object: object with global id %s not found" 
+					% arguments[0]
```
```suggestion
		escoria.logger.error(
			self,
			get_command_name() + ": The object to turn towards with global id %s was not found." 
					% arguments[0]
```

balloonpopper commented on addons/escoria-core/game/core-scripts/esc/commands/turn_to.gd:
```diff
@@ -65,9 +62,7 @@ func run(command_params: Array) -> int:
-		[
-			"Interrupt() function not implemented"
-		]
+	escoria.logger.warn(
+		self,
+		"[%s] Interrupt() function not implemented" % get_command_name()
```
```suggestion
	escoria.logger.warn(
		self,
		"[%s] Interrupt() function not implemented." % get_command_name()
```

balloonpopper commented on addons/escoria-core/game/core-scripts/esc/commands/walk.gd:
```diff
@@ -38,19 +38,17 @@ func validate(arguments: Array):
-			[
-				"Object with global id %s not found" % arguments[0]
-			]
+		escoria.logger.error(
+			self,
+			get_command_name() + ": invalid first object. " +
+			"Object with global id %s not found" % arguments[0]
```
```suggestion
		escoria.logger.error(
			self,
			get_command_name() + ": invalid first object. " +
			"The object with global id %s to make walk was not found." % arguments[0]
```

balloonpopper commented on addons/escoria-core/game/core-scripts/esc/commands/walk.gd:
```diff
@@ -38,19 +38,17 @@ func validate(arguments: Array):
-			[
-				"Object with global id %s not found" % arguments[0]
-			]
+		escoria.logger.error(
+			self,
+			get_command_name() + ": invalid second object. " +
+			"Object with global id %s not found" % arguments[0]
```
```suggestion
		escoria.logger.error(
			self,
			get_command_name() + ": invalid second object. " +
			"The object to walk to with global id %s was not found." % arguments[1]
```

balloonpopper commented on addons/escoria-core/game/core-scripts/esc/commands/walk_block.gd:
```diff
@@ -38,19 +38,17 @@ func validate(arguments: Array):
-			[
-				"Object with global id %s not found" % arguments[0]
-			]
+		escoria.logger.error(
+			self,
+			get_command_name() + ": invalid first object. " + 
+			"Object with global id %s not found" % arguments[0]
```
```suggestion
		escoria.logger.error(
			self,
			get_command_name() + ": invalid first object. " + 
			"The object to make walk with global id %s was not found." % arguments[0]
```

balloonpopper commented on addons/escoria-core/game/core-scripts/esc/commands/walk_block.gd:
```diff
@@ -38,19 +38,17 @@ func validate(arguments: Array):
-			[
-				"Object with global id %s not found" % arguments[1]
-			]
+		escoria.logger.error(
+			self,
+			get_command_name() + ": invalid second object. " +
+			"Object with global id %s not found" % arguments[1]
```
```suggestion
		escoria.logger.error(
			self,
			get_command_name() + ": invalid second object. " +
			"The object to walk to with global id %s was not found." % arguments[1]
```

balloonpopper commented on addons/escoria-core/game/core-scripts/esc/commands/walk_to_pos.gd:
```diff
@@ -37,11 +37,10 @@ func validate(arguments: Array):
-			[
-				"Object with global id %s not found" % arguments[0]
-			]
+		escoria.logger.error(
+			self,
+			get_command_name() + ": invalid first object. " +
+			"Object with global id %s not found" % arguments[0]
```
```suggestion
		escoria.logger.error(
			self,
			get_command_name() + ": invalid first object. " +
			"The object to make walk with global id %s was not found." % arguments[0]
```

balloonpopper commented on addons/escoria-core/game/core-scripts/esc/commands/walk_to_pos_block.gd:
```diff
@@ -37,11 +37,10 @@ func validate(arguments: Array):
-			[
-				"Object with global id %s not found" % arguments[0]
-			]
+		escoria.logger.error(
+			self,
+			get_command_name() + ": invalid first object. " +
+					"Object with global id %s not found" % arguments[0]
```
```suggestion
		escoria.logger.error(
			self,
			get_command_name() + ": invalid first object. " +
					"The object to make walk with global id %s was not found." % arguments[0]
```

balloonpopper commented on addons/escoria-core/game/core-scripts/save_data/esc_save_manager.gd:
```diff
@@ -216,14 +216,16 @@ func load_game(id: int):
-		escoria.logger.report_errors(
-			"esc_save_manager.gd:load_game()",
-			["Save file %s doesn't exist" % save_file_path])
+		escoria.logger.error(
+			self,
+			"Save file %s doesn't exist" % save_file_path
+		)
```
```suggestion
		escoria.logger.error(
			self,
			"Save file %s doesn't exist." % save_file_path
		)
```

balloonpopper commented on addons/escoria-core/game/core-scripts/save_data/esc_save_manager.gd:
```diff
@@ -216,14 +216,16 @@ func load_game(id: int):
 	escoria.logger.info(
-		"esc_save_manager.gd:load_game()",
-		["Loading savegame %s" % str(id)])
+		self,
+		"Loading savegame %s" % str(id)
+	)
```
```suggestion
		self,
		"Loading savegame %s." % str(id)
	)
```

balloonpopper commented on addons/escoria-core/game/core-scripts/save_data/esc_save_manager.gd:
```diff
@@ -432,7 +432,7 @@ func save_settings():
 	var error: int = ResourceSaver.save(save_path, settings_res)
 	if error != OK:
-		escoria.logger.report_errors(
+		escoria.logger.error(
 			"esc_save_manager.gd:save_settings()",
 			["There was an issue writing settings %s" % save_path])
```
```suggestion
		escoria.logger.error(
			"esc_save_manager.gd:save_settings()",
			["There was an issue writing settings file %s." % save_path])
```

balloonpopper commented on addons/escoria-core/game/esc_autoload.gd:
```diff
@@ -0,0 +1,184 @@
+var is_direct_room_run: bool = false
+
+# Ready function
+func _ready():
+	# We check if we run the full game or directly a room scene
+	if not get_tree().current_scene is ESCMain:
+		# Running a room scene. We need to instanciate the main scene ourselves
+		# so that the Escoria scene is created and managers are instanced as well.
```
```suggestion
func _ready():
	# We check if we run the full game or a room scene directly
	if not get_tree().current_scene is ESCMain:
		# Running a room scene. We need to instantiate the main scene ourselves
		# so that the Escoria scene is created and managers are instanced as well.
```

balloonpopper commented on addons/escoria-core/game/esc_autoload.gd:
```diff
@@ -0,0 +1,184 @@
+		add_child(main_scene)
+
+
+# Get the Escoria node. That node gives access to 
```
The comment is missing the end of the sentence

StraToN replied:
Fixed

balloonpopper commented on addons/escoria-core/game/esc_autoload.gd:
```diff
@@ -0,0 +1,184 @@
+
+# Get the Escoria node. That node gives access to 
+func get_escoria():
+	# We check if we run the full game or directly a room scene
```
```suggestion
	# We check if we run the full game or a room scene directly
```

balloonpopper commented on addons/escoria-core/game/esc_inputs_manager.gd:
```diff
@@ -166,7 +166,10 @@ func clear_stack():
 func _on_left_click_on_bg(position: Vector2) -> void:
 	if input_mode == INPUT_ALL and hotspot_focused.empty():
-		escoria.logger.info("Left click on background at ", [str(position)])
+		escoria.logger.info(
+			self,
+			"Left click on background at %s" % str(position)
+		)
```
```suggestion
		escoria.logger.info(
			self,
			"Left click on background at %s." % str(position)
		)
```

balloonpopper commented on addons/escoria-core/game/esc_inputs_manager.gd:
```diff
@@ -178,6 +181,7 @@ func _on_left_click_on_bg(position: Vector2) -> void:
 func _on_double_left_click_on_bg(position: Vector2) -> void:
 	if input_mode == INPUT_ALL and hotspot_focused.empty():
 		escoria.logger.info(
+			self,
 			"Double left click on background at %s" % str(position)
```
```suggestion
		escoria.logger.info(
			self,
			"Double left click on background at %s." % str(position)
```

balloonpopper commented on addons/escoria-core/game/esc_inputs_manager.gd:
```diff
@@ -190,7 +194,10 @@ func _on_double_left_click_on_bg(position: Vector2) -> void:
 func _on_right_click_on_bg(position: Vector2) -> void:
 	if input_mode == INPUT_ALL and hotspot_focused.empty():
-		escoria.logger.info("Right click on background at ", [str(position)])
+		escoria.logger.info(
+			self,
+			"Right click on background at %s " % str(position)
+		)
```
```suggestion
		escoria.logger.info(
			self,
			"Right click on background at %s." % str(position)
		)
```

balloonpopper commented on addons/escoria-core/game/esc_inputs_manager.gd:
```diff
@@ -205,6 +212,7 @@ func _on_mouse_left_click_inventory_item(
 	event: InputEvent
 ) -> void:
 	escoria.logger.info(
+		self,
 		"Inventory item left clicked %s " % inventory_item_global_id
```
```suggestion
	escoria.logger.info(
		self,
		"Inventory item %s left clicked." % inventory_item_global_id
```

balloonpopper commented on addons/escoria-core/game/esc_inputs_manager.gd:
```diff
@@ -225,6 +233,7 @@ func _on_mouse_right_click_inventory_item(
 ) -> void:
 	if input_mode == INPUT_ALL:
 		escoria.logger.info(
+			self,
 			"Inventory item right clicked %s" % inventory_item_global_id
```
```suggestion
		escoria.logger.info(
			self,
			"Inventory item %s right clicked." % inventory_item_global_id
```

balloonpopper commented on addons/escoria-core/game/esc_inputs_manager.gd:
```diff
@@ -245,6 +254,7 @@ func _on_mouse_double_left_click_inventory_item(
 ) -> void:
 	if input_mode == INPUT_ALL:
 		escoria.logger.info(
+			self,
 			"Inventory item double left clicked %s" % inventory_item_global_id
```
```suggestion
		escoria.logger.info(
			self,
			"Inventory item %s double left clicked." % inventory_item_global_id
```

balloonpopper commented on addons/escoria-core/game/esc_inputs_manager.gd:
```diff
@@ -261,6 +271,7 @@ func _on_mouse_double_left_click_inventory_item(
 #	that is hovered
 func _on_mouse_entered_inventory_item(inventory_item_global_id: String) -> void:
 	escoria.logger.info(
+		self,
 		"Inventory item focused %s" % inventory_item_global_id
```
```suggestion
	escoria.logger.info(
		self,
		"Inventory item %s focused." % inventory_item_global_id
```

balloonpopper commented on addons/escoria-core/game/esc_inputs_manager.gd:
```diff
@@ -270,7 +281,10 @@ func _on_mouse_entered_inventory_item(inventory_item_global_id: String) -> void:
 # The mouse exited an inventory item
 func _on_mouse_exited_inventory_item() -> void:
-	escoria.logger.info("Inventory item unfocused")
+	escoria.logger.info(
+		self,
+		"Inventory item unfocused"
+	)
```
```suggestion
	escoria.logger.info(
		self,
		"Inventory item unfocused."
	)
```

balloonpopper commented on addons/escoria-core/game/esc_inputs_manager.gd:
```diff
@@ -320,7 +340,10 @@ func _on_mouse_exited_item(item: ESCItem) -> void:
 	if input_mode == INPUT_ALL:
 		if hover_stack.empty() or hover_stack.back() == item:
-			escoria.logger.info("Item left clicked", [item.global_id, event])
+			escoria.logger.info(
+				self,
+				"Item left clicked: %s with event %s" % [item.global_id, event]
+			)
```
```suggestion
			escoria.logger.info(
				self,
				"Item %s left clicked with event %s." % [item.global_id, event]
			)
```

balloonpopper commented on addons/escoria-core/game/esc_inputs_manager.gd:
```diff
@@ -339,7 +362,10 @@ func _on_mouse_left_double_clicked_item(
 ) -> void:
 	if input_mode == INPUT_ALL:
-		escoria.logger.info("Item left double clicked", [item.global_id, event])
+		escoria.logger.info(
+			self,
+			"Item left double clicked %s with event %s" % [item.global_id, event]
+		)
```
```suggestion
		escoria.logger.info(
			self,
			"Item %s left double clicked with event %s." % [item.global_id, event]
		)
```

balloonpopper commented on addons/escoria-core/game/esc_inputs_manager.gd:
```diff
@@ -355,7 +381,10 @@ func _on_mouse_left_double_clicked_item(
 func _on_mouse_right_clicked_item(item: ESCItem, event: InputEvent) -> void:
 	if input_mode == INPUT_ALL:
-		escoria.logger.info("Item right clicked", [item.global_id, event])
+		escoria.logger.info(
+			self,
+			"Item right clicked %s with event %s" % [item.global_id, event]
+		)
```
```suggestion
		escoria.logger.info(
			self,
			"Item %s right clicked with event %s." % [item.global_id, event]
		)
```

balloonpopper commented on addons/escoria-core/game/esc_logger.gd:
```diff
@@ -0,0 +1,193 @@
+		# Open logfile in write mode
+		log_file = File.new()
+
+		# this is left alone as this constructor is called from escoria.gd's own
+		# constructor
```
```suggestion
		# This is left alone as this constructor is called from escoria.gd's own
		# constructor
```

balloonpopper commented on addons/escoria-core/game/main.gd:
```diff
@@ -38,7 +38,10 @@ func _ready() -> void:
 func set_scene(p_scene: Node) -> void:
 	if !p_scene:
-		escoria.logger.report_errors("main", ["Trying to set empty scene"])
+		escoria.logger.error(
+			self, 
+			"Trying to set empty scene"
+		)
```
Please confirm whether I described what this error is saying or not.
```suggestion
		escoria.logger.error(
			self, 
			"Can't change to an empty scene. Please specify the scene name."
		)
```

balloonpopper commented on addons/escoria-core/game/main.gd:
```diff
@@ -151,15 +152,16 @@ func set_camera_limits(camera_limit_id: int = 0, scene: Node = current_scene) ->
-				[
-					"No limit area! Using viewport."
-				]
+			escoria.logger.warn(
+				self,
+				"No limit area! Using viewport."
```
```suggestion
			escoria.logger.warn(
				self,
				"Defined camera is smaller than the viewport. Using viewport size."
```

balloonpopper commented on addons/escoria-core/game/main.gd:
```diff
@@ -176,8 +178,8 @@ func set_camera_limits(camera_limit_id: int = 0, scene: Node = current_scene) ->
 		escoria.logger.info(
-			"Setting camera limits from parameter ",
-			[scene_camera_limits]
+			self,
+			"Setting camera limits from parameter " + str(scene_camera_limits)
```
```suggestion
			self,
			"Setting camera limits using configured parameters " + str(scene_camera_limits)
```

balloonpopper commented on addons/escoria-core/game/scenes/camera_player/esc_camera.gd:
```diff
@@ -103,20 +103,19 @@ func set_target(p_target, p_time : float = 0.0):
 	_resolve_target_and_zoom(p_target)
 
 	escoria.logger.info(
+		self,
 		"Current camera position = %s " % str(self.global_position)
```
```suggestion
		self,
		"Current camera position = %s." % str(self.global_position)
```

balloonpopper commented on addons/escoria-core/game/scenes/camera_player/esc_camera.gd:
```diff
@@ -103,20 +103,19 @@ func set_target(p_target, p_time : float = 0.0):
-						_tween.tell(),
-						_tween.get_runtime()
-					]
+			escoria.logger.warn(
+				self,
+				"Tween is still active: %f/%f" % [
+					_tween.tell(),
+					_tween.get_runtime()
```
```suggestion
			escoria.logger.warn(
				self,
				"Tween is still active: %f seconds of %f completed." % [
					_tween.tell(),
					_tween.get_runtime()
```

balloonpopper commented on addons/escoria-core/game/scenes/camera_player/esc_camera.gd:
```diff
@@ -140,9 +139,9 @@ func set_target(p_target, p_time : float = 0.0):
-		escoria.logger.report_errors(
-			"esc_camera.gd:set_camera_zoom()",
-			["Tried to set negative or zero zoom level"]
+		escoria.logger.error(
+			self,
+			"Tried to set negative or zero zoom level"
```
```suggestion
		escoria.logger.error(
			self,
			"Tried to set negative or zero zoom level."
```

balloonpopper commented on addons/escoria-core/game/scenes/camera_player/esc_camera.gd:
```diff
@@ -199,13 +196,11 @@ func push(p_target, p_time: float = 0.0, p_type: int = 0):
-						_tween.tell(),
-						_tween.get_runtime()
-					]
+			escoria.logger.warn(
+				self,
+				"Tween is still active: %f/%f" % [
+					_tween.tell(),
+					_tween.get_runtime()
```
```suggestion
			escoria.logger.warn(
				self,
				"Tween is still active: %f seconds of %f completed." % [
					_tween.tell(),
					_tween.get_runtime()
```

balloonpopper commented on addons/escoria-core/game/scenes/camera_player/esc_camera.gd:
```diff
@@ -251,15 +246,13 @@ func shift(p_target: Vector2, p_time: float, p_type: int):
-					]
-				]
-			)
+		escoria.logger.warn(
+			self,
+			"Tween is still active: %f/%f" % [
+				_tween.tell(),
+				_tween.get_runtime()
+			]
+		)
```
```suggestion
		escoria.logger.warn(
			self,
			"Tween is still active: %f seconds of %f completed." % [
				_tween.tell(),
				_tween.get_runtime()
			]
		)
```

balloonpopper commented on addons/escoria-core/game/scenes/dialogs/esc_dialog_options_chooser.gd:
```diff
@@ -26,9 +26,15 @@ func set_dialog(new_dialog: ESCDialog) -> void:
 # Show the dialog chooser UI
 func show_chooser() -> void:
-	escoria.logger.error("Dialog chooser did not implement the show method.")
+	escoria.logger.error(
+		self,
+		"Dialog chooser did not implement the show method."
+	)
```
```suggestion
	escoria.logger.error(
		self,
		"Dialog chooser does not implement the show method."
	)
```

balloonpopper commented on addons/escoria-core/game/scenes/dialogs/esc_dialog_options_chooser.gd:
```diff
@@ -26,9 +26,15 @@ func set_dialog(new_dialog: ESCDialog) -> void:
 # Hide the dialog chooser UI
 func hide_chooser() -> void:
-	escoria.logger.error("Dialog chooser did not implement the hide method.")
+	escoria.logger.error(
+		self,
+		"Dialog chooser did not implement the hide method."
+	)
```
```suggestion
	escoria.logger.error(
		self,
		"Dialog chooser does not implement the hide method."
	)
```

balloonpopper commented on addons/escoria-core/game/scenes/dialogs/esc_dialog_player.gd:
```diff
@@ -109,39 +109,33 @@ func say(character: String, type: String, text: String) -> void:
-			[
-				"No dialog manager supports the type %s" % type
-			]
+		escoria.logger.error(
+			self,
+			"No dialog manager supports the type %s" % type
```
Please confirm I've understood the error message correctly.
```suggestion
		escoria.logger.error(
			self,
			"No dialog manager called %s configured." % type
```

balloonpopper commented on addons/escoria-core/game/scenes/dialogs/esc_dialog_player.gd:
```diff
@@ -109,39 +109,33 @@ func say(character: String, type: String, text: String) -> void:
-			[
-				"Unexpected text encountered %s" % text
-			]
+		escoria.logger.error(
+			self,
+			"Unexpected text encountered %s" % text
```
```suggestion
		escoria.logger.error(
			self,
			"Unexpected text encountered : %s." % text
```

balloonpopper commented on addons/escoria-core/game/scenes/dialogs/esc_dialog_player.gd:
```diff
@@ -164,27 +158,25 @@ func speedup() -> void:
-		escoria.logger.report_errors(
-			"esc_dialog_player.gd:start_dialog_choices()",
-			["Received answers array was empty."]
+		escoria.logger.error(
+			self,
+			"Received answers array was empty."
```
```suggestion
		escoria.logger.error(
			self,
			"Received dialog options array was empty."
```

balloonpopper commented on addons/escoria-core/game/scenes/dialogs/esc_dialog_player.gd:
```diff
@@ -164,27 +158,25 @@ func speedup() -> void:
-			[
-				"No dialog manager supports the chooser type %s" % type
-			]
+		escoria.logger.error(
+			self,
+			"No dialog manager supports the chooser type %s" % type
```
```suggestion
		escoria.logger.error(
			self,
			"No dialog manager supports the chooser type %s." % type
```

balloonpopper commented on addons/escoria-core/game/scenes/inventory/inventory_ui.gd:
```diff
@@ -17,9 +17,9 @@ var items_ids_in_inventory: Dictionary = {}
-		escoria.logger.report_errors(
-			self.get_path(),
-			["Items container is empty."]
+		escoria.logger.error(
+			self,
+			"Items container is empty."
```
Can we add some helper text with how to resolve this please? 
```suggestion
		escoria.logger.error(
			self,
			"Inventory items container is empty."
```

balloonpopper commented on addons/escoria-core/game/scenes/inventory/inventory_ui.gd:
```diff
@@ -158,10 +156,7 @@ func _on_escoria_global_changed(global: String, old_value, new_value) -> void:
-				"Global must contain only one item name.",
-				"(received: %s)" % global
-			]
+		escoria.logger.error(
+			self,
+			"Global must contain only one item name (received: %s)" % global
```
```suggestion
		escoria.logger.error(
			self,
			"Global must contain only one item name (received: %s)." % global
```

balloonpopper commented on addons/escoria-ui-9verbs/game.gd:
```diff
@@ -363,8 +363,9 @@ func _on_event_done(_return_code: int, _event_name: String):
 		escoria.logger.info(
-			"custom setting value loaded:",
-			[custom_settings["a_custom_setting"]]
+			self,
+			"custom setting value loaded: %s" 
+					% str(custom_settings["a_custom_setting"])
```
```suggestion
			self,
			"custom setting value loaded: %s." 
					% str(custom_settings["a_custom_setting"])
```

balloonpopper commented on addons/escoria-ui-9verbs/plugin.gd:
```diff
@@ -3,16 +3,13 @@ tool
 # Deregister UI
-func _exit_tree() -> void:
-	escoria.deregister_ui("res://addons/escoria-ui-9verbs/game.tscn")
+func disable_plugin():
+	print("Disabling plugin Escoria UI 9-verbs")
+	EscoriaPlugin.deregister_ui("res://addons/escoria-ui-9verbs/game.tscn")
```
```suggestion
func disable_plugin():
	print("Disabling plugin Escoria UI 9-verbs.")
	EscoriaPlugin.deregister_ui("res://addons/escoria-ui-9verbs/game.tscn")
```

balloonpopper commented on addons/escoria-ui-9verbs/plugin.gd:
```diff
@@ -3,16 +3,13 @@ tool
 # Register UI with Escoria
-func _register():
-	escoria.register_ui("res://addons/escoria-ui-9verbs/game.tscn")
+func enable_plugin():
+	print("Enabling plugin Escoria UI 9-verbs")
+	EscoriaPlugin.register_ui("res://addons/escoria-ui-9verbs/game.tscn")
```
```suggestion
func enable_plugin():
	print("Enabling plugin Escoria UI 9-verbs")
	EscoriaPlugin.register_ui("res://addons/escoria-ui-9verbs/game.tscn")
```
```suggestion
func enable_plugin():
	print("Enabling plugin Escoria UI 9-verbs.")
	EscoriaPlugin.register_ui("res://addons/escoria-ui-9verbs/game.tscn")
```

balloonpopper commented on addons/escoria-ui-keyboard-9verbs/plugin.gd:
```diff
@@ -1,18 +1,15 @@
 # Deregister UI
-func _exit_tree() -> void:
-	escoria.deregister_ui("res://addons/escoria-ui-keyboard-9verbs/game.tscn")
+func disable_plugin() -> void:
+	print("Disabling plugin Escoria UI 9-verbs with keyboard")
+	EscoriaPlugin.deregister_ui("res://addons/escoria-ui-keyboard-9verbs/game.tscn")
```
```suggestion
func disable_plugin() -> void:
	print("Disabling plugin Escoria UI 9-verbs with keyboard.")
	EscoriaPlugin.deregister_ui("res://addons/escoria-ui-keyboard-9verbs/game.tscn")
```

balloonpopper commented on addons/escoria-ui-keyboard-9verbs/plugin.gd:
```diff
@@ -1,18 +1,15 @@
 # Register UI with Escoria
-func _register():
-	escoria.register_ui("res://addons/escoria-ui-keyboard-9verbs/game.tscn")
+func enable_plugin():
+	print("Enabling plugin Escoria UI 9-verbs with keyboard")
+	EscoriaPlugin.register_ui("res://addons/escoria-ui-keyboard-9verbs/game.tscn")
```
```suggestion
func enable_plugin():
	print("Enabling plugin Escoria UI 9-verbs with keyboard.")
	EscoriaPlugin.register_ui("res://addons/escoria-ui-keyboard-9verbs/game.tscn")
```

balloonpopper commented on addons/escoria-ui-simplemouse/game.gd:
```diff
@@ -328,8 +328,9 @@ func pause_game():
 		escoria.logger.info(
-			"custom setting value loaded:",
-			[custom_settings["a_custom_setting"]]
+			self,
+			"custom setting value loaded: %s" 
+					% str(custom_settings["a_custom_setting"])
```
```suggestion
			self,
			"custom setting value loaded: %s." 
					% str(custom_settings["a_custom_setting"])
```

balloonpopper commented on addons/escoria-ui-simplemouse/plugin.gd:
```diff
@@ -3,16 +3,13 @@ tool
 # Deregister UI
-func _exit_tree() -> void:
-	escoria.deregister_ui("res://addons/escoria-ui-simplemouse/game.tscn")
+func disable_plugin():
+	print("Disabling plugin Escoria UI Simple Mouse")
```
```suggestion
func disable_plugin():
	print("Disabling plugin Escoria UI Simple Mouse.")
```

balloonpopper commented on addons/escoria-ui-simplemouse/plugin.gd:
```diff
@@ -3,16 +3,13 @@ tool
 # Register UI with Escoria
-func _register():
-	escoria.register_ui("res://addons/escoria-ui-simplemouse/game.tscn")
+func enable_plugin():
+	print("Enabling plugin Escoria UI Simple Mouse")
```
```suggestion
func enable_plugin():
	print("Enabling plugin Escoria UI Simple Mouse.")
```

balloonpopper commented on addons/escoria-core/game/escoria.gd:
```diff
@@ -246,121 +104,27 @@ func set_game_paused(p_paused: bool):
-			"escoria.gd:run_event_from_script()",
-			["Requested action %s on unloaded script %s" % [event_name, script],
-			"Please load the ESC script using esc_compiler.load_esc_file()."]
+		escoria.logger.error(
+			self,
+			"Requested action %s on unloaded script %s" % [event_name, script] +
+			"Please load the ESC script using esc_compiler.load_esc_file()."
```
```suggestion
		escoria.logger.error(
			self,
			"Requested action %s on unloaded script %s." % [event_name, script] +
			"Please load the ESC script using esc_compiler.load_esc_file()."
```

balloonpopper commented on addons/escoria-core/game/escoria.gd:
```diff
@@ -246,121 +104,27 @@ func set_game_paused(p_paused: bool):
-		self.logger.report_errors(
-			"Start event of the start script returned unsuccessful: %d" % rc[0],
-			[]
+		escoria.logger.error(
+			self,
+			"Start event of the start script returned unsuccessful: %d" % rc[0]
```
```suggestion
		escoria.logger.error(
			self,
			"Start event of the start script returned unsuccessful: %d." % rc[0]
```

balloonpopper commented on addons/escoria-core/ui_library/tools/room_select/room_select.gd:
```diff
@@ -44,8 +44,10 @@ func _ready():
 	else:
-		escoria.logger.report_warnings("room_select.gd:_ready()",
-			["A problem occurred while opening rooms folder."])
+		escoria.logger.warn(
+			self,
+			"A problem occurred while opening rooms folder."
+		)
```
```suggestion
		escoria.logger.warn(
			self,
			"A problem occurred while opening rooms folder %s." % str(path)
		)
```

balloonpopper commented:
Issues encountered when testing: 
* Crash going from room 9 to room 8. 
To reproduce it you have to start in level 8, open the door to room 9, then go back to room 8. It does not occur if you start on room 9 and go to room 8.
Also happens if you start in room 9, go to 10, then back to 9.
Doesn't happen between rooms 1 and 2 no matter how many times I switch... 🤷
* It also crashes after you interact with something. If I go 1-2-1-2-1-2-1 it doesn't trigger, but if I go into room 2 and press the 1st button, then go to room 1 then try to go to room 2 it crashes.
* Pick up the pen and paper in room 5. Use the pen with the paper -> crash. 
* Room 10 - Use the "accept input test" he walks left, then right, crash. 
* Every time I adjust the volume, this prints in the output window
"ESC (2022-05-15T21:59:18) I game.gd: custom setting value loaded: 100
ESC (2022-05-15T21:59:18) I esc_autoload.gd: *** settings loaded"
* Start in room 9. Move a bit. Save. Move a bit more in room 9. Load game. Crash.
* If I start a full game and then choose room 1, room 1 transitions in and the music starts as you'd expect. If I quit and just run room 1 directly, the music still starts. If I start any room directly other than room 1, no music.
* double right click a picture in room 1 = crash
* Right hand picture in room 1 is missing the talking head when you look at the picture. You still get the box with the text, but the talking head is missing.

StraToN commented on addons/escoria-core/game/core-scripts/esc/esc_object_manager.gd:
```diff
@@ -69,13 +69,13 @@ func _init() -> void:
-			for object in room.objects:
-				if (object as ESCObject).node:
-					(object as ESCObject).node.visible = (object as ESCObject).active
+## Make active objects in current room visible
+#func _process(_delta):
+#	for room in room_objects:
+#		if room.is_reserved or _is_current_room(room):
+#			for object in room.objects:
+#				if (object as ESCObject).node:
+#					(object as ESCObject).node.visible = (object as ESCObject).active
```
Oops I forgot to retest this. I wanted to verify if that function was really necessary. I felt it was quite heavy to use a _process() function with a for-loop on all objects in the room for the purpose of showing/hiding them if necessary.
Not resolved, to be tested

StraToN replied:
Fixed. These lines are now removed as they seem not necessary. @BHSDuncan 
Reserved objects do not need to be "visible", and other objects' `ESCObject._set_active()` function is automatically called to set them active and visible/invisible at the same time.

StraToN commented on addons/escoria-core/game/core-scripts/esc/commands/accept_input.gd:
```diff
@@ -37,12 +37,11 @@ func validate(arguments: Array):
-		escoria.logger.report_errors(
-			"accept_input: invalid parameter",
-			[
+		escoria.logger.error(
+			self,
+			get_command_name() + ": invalid parameter. " +
 				"%s is not a valid parameter value (ALL, NONE, SKIP)" %\
 						arguments[0]
-			]
 		)
```
Good idea. Done.

StraToN commented on addons/escoria-core/game/esc_logger.gd:
```diff
@@ -0,0 +1,193 @@
+
+	func formatted_message(context: String, msg: String, letter: String) -> String:
+		return "ESC ({0}) {1} {2}: {3}".format([letter, context, _formatted_date(), msg])
+
```
Done

StraToN commented:
> * Crash going from room 9 to room 8.
>       To reproduce it you have to start in level 8, open the door to room 9, then go back to room 8. It does not occur if you start on room 9 and go to room 8.
>       Also happens if you start in room 9, go to 10, then back to 9.
>       Doesn't happen between rooms 1 and 2 no matter how many times I switch... shrug

I retested this issue AFTER commit [9380731](https://github.com/godot-escoria/escoria-demo-game/pull/598/commits/9380731a71d2fd018a7187c1c335e8452df767e9) and was not able to reproduce. I assume the issue was caused by the fact that when you enter a room that contains an item which state is *not* `default`, then this state is saved in the object manager. When exiting the room, this object is kept in the Object Manager (under its room key). However, re-entering the room indeed crashed on [addons/escoria-core/game/core-scripts/esc/esc_object_manager.gd](https://github.com/godot-escoria/escoria-demo-game/blob/4f77309e5d414eeb2390f6c161058f47e5e0e839/addons/escoria-core/game/core-scripts/esc/esc_object_manager.gd#L154) precisely because re-entering the room tries to register the item that was previously kept in the Object Manager because of its state being different from `default`.

I added a condition for this error: the saved object state must be `default` for the error to trigger. 


> It also crashes after you interact with something. If I go 1-2-1-2-1-2-1 it doesn't trigger, but if I go into room 2 and press the 1st button, then go to room 1 then try to go to room 2 it crashes.

Same bug as above. Fixed.


> Pick up the pen and paper in room 5. Use the pen with the paper -> crash.

This was caused by a wrong setting in Project Settings: `escoria/ui/items_autoregister_path` was set to `res://game/items/escitems/` but that folder only contains the button scene, not inventory items. Right folder was `res://game/items/inventory/`. Fixed.

> Room 10 - Use the "accept input test" he walks left, then right, crash.

Wrong dialog type set in file `res://game/rooms/room10/esc/button_accept_input_test.esc` L9:
```
say player "Ha! Now you can't even skip this text!" avatar_dialog_player
```
Changed to:
```
say player "Ha! Now you can't even skip this text!" avatar
```
Fixes the issue.

> Every time I adjust the volume, this prints in the output window
> "ESC (2022-05-15T21:59:18) I game.gd: custom setting value loaded: 100
> ESC (2022-05-15T21:59:18) I esc_autoload.gd: *** settings loaded"

This is normal. When any sound setting is changed (ie. cursor is dragged), the `esc_autoload.gd:apply_settings()` function is called to apply the new setting. This function actually applies ALL settings at the same time, not just the one that changed. This also calls the `game_scene.apply_custom_settings()` function that loads a custom setting and logs it.

> Start in room 9. Move a bit. Save. Move a bit more in room 9. Load game. Crash.

Saving and loading is still broken at the moment as I didn't have the time to work on it yet. Not related to this PR. Will be fixed when I'll fix https://github.com/godot-escoria/escoria-issues/issues/217 (this is my upcoming task after this PR is merged).


> If I start a full game and then choose room 1, room 1 transitions in and the music starts as you'd expect. If I quit and just run room 1 directly, the music still starts. If I start any room directly other than room 1, no music.

Music is set by room1's ESC script's `:setup` event so I assume this is working as expected.

> double right click a picture in room 1 = crash

Could not reproduce. Maybe fixed somehow?

> Right hand picture in room 1 is missing the talking head when you look at the picture. You still get the box with the text, but the talking head is missing.

The Avatars path was missing in project settings. Fixed.

balloonpopper commented:
@StraToN Generally everything is working (great job!) and it's a lot more stable than the last test. The issues I spotted this time are below. Some look to me like they aren't related to this PR, feel free to say they're not related and I'll raise them separately.

* Room6 - Where did the graphics and logic go? They've reverted back to the old version
* In room5, using the "pen" on the "pipe" crashes. (esc_action_manager.gd line 326)
```
"Invalid action" + errors
```
If I change that to 
```
"Invalid action" + str(errors)
```
then the error goes away. However, while doing this makes the warning show up in the Godot console, the warning message no longer appears in the debugger console (as it now uses "push_warning")
Can we get the print statement added for warning and errors please?
```
print(formatted_message(context, msg, "W"))
<...>
print(formatted_message(context, msg, "E"))
```
* Lots of errors get logged when you use the pen with the paper
![image](https://user-images.githubusercontent.com/5151242/168935018-715643d1-8499-4b0f-8413-5b2543e623de.png)
* I noticed when walking that double clicking is inconsistent. I can't reliably reproduce it, but if you do lots of double clicking around the room, it feels like ~25% of the time the double click doesn't work when it should. If you try to make him run right-left-right-left-etc you'll notice that sometimes the double click is registered as 2 single clicks.
* Room 11 crashes on exit (11->12) as does 13->12. In room 12, the ESCLocation for the left door has the same global_id as the left door parent node. The room works when that's fixed.
* Room 11 buttons don't work (reasonably sure they do in develop?)
* Most of the buttons in room 14 don't work (I'm pretty sure they're working in develop branch)
* If you go room 12->13 then RIGHT click the left door you can't then left click the door to exit. The "Use" action isn't cleared after a use action fails, so use is stuck turned on. I noticed used wasn't cleared when you do an invalid action in room 5 too (use pen with pipe for example).
* If I change the "Command directories" setting to an invalid folder, it (correctly) tells you
```
ESC (2022-05-18T11:40:29) E esc_compiler.gd: Invalid command detected: change_scene
Command change_scene is not recognised.
```
I was thinking we might want to make that message a bit more helpful. Something like
"Command change_scene cannot be found under folder %s. Please confirm setting "escoria/main/command directories" is set to the folder where the ESC commands are found."
* Setting "text lang" and "voice lang" to invalid locales doesn't print an error, but you don't get any text when you look at the paintings in room 1.
* After I'd tested the above by setting the language settings to "fr_fr", I re-set them back to "fr_FR" and played room 1 again. The text is still broken and doesn't appear when you look at either painting in room 1. (I ended up having to close Godot and git revert everything then restart Godot to get text back)
* If you look at painting 1 and while he's halfway through "I don't know what that stuff is" you hit Escape and change languages, he finishes saying the line in the original language. The next time you look at the painting, Godot locks up.
* 

Retesting previous bugs 
* Not quite a previous

balloonpopper commented on addons/escoria-core/game/core-scripts/esc_item.gd:
```diff
@@ -367,8 +370,9 @@ func get_animation_player() -> Node:
-				"Can not find node at path %s" % player_node_path
+			escoria.logger.warn(
+				self,
+				"Can not find animation_player node at path %s." % player_node_path
```
Based off the error I just reported
```suggestion
				"Can not find animation_player node at path %s for %s." % [player_node_path, global_id]
```

balloonpopper commented on addons/escoria-core/game/core-scripts/esc/esc_action_manager.gd:
```diff
@@ -314,16 +321,16 @@ func _activate(
-							escoria.logger.report_warnings(
-								"ESCActionManager.activate: Invalid action",
-								errors
+							escoria.logger.warn(
+								self,
+								"Invalid action" + errors
 							)
```
```suggestion
							escoria.logger.warn(
								self,
								"Invalid action" + str(errors)
							)
```

balloonpopper commented on addons/escoria-core/game/esc_logger.gd:
```diff
@@ -0,0 +1,193 @@
+		print(formatted_message(context, msg, "I"))
+
+
+	# Warning log
+	func warn(owner: Object, msg: String):
+		var context = owner.get_script().resource_path.get_file()
+		push_warning(formatted_message(context, msg, "W"))
+		if ESCProjectSettingsManager.get_setting(
+			ESCProjectSettingsManager.TERMINATE_ON_WARNINGS
+		):
+			assert(false)
+			escoria.get_tree().quit()
+			
+	
+	# Error log
+	func error(owner: Object, msg: String):
+		var context = owner.get_script().resource_path.get_file()
+		push_error(formatted_message(context, msg, "E"))
+		if ESCProjectSettingsManager.get_setting(
+			ESCProjectSettingsManager.TERMINATE_ON_ERRORS
+		):
+			assert(false)
+			escoria.get_tree().quit()
```
Suggest we add these lines as per below
print(formatted_message(context, msg, "W"))
print(formatted_message(context, msg, "E"))
as currently warnings and errors don't print in the debugger output, only the Godot console.

```suggestion
	# Warning log
	func warn(owner: Object, msg: String):
		var context = owner.get_script().resource_path.get_file()
		print(formatted_message(context, msg, "W"))
		push_warning(formatted_message(context, msg, "W"))
		if ESCProjectSettingsManager.get_setting(
			ESCProjectSettingsManager.TERMINATE_ON_WARNINGS
		):
			assert(false)
			escoria.get_tree().quit()
			
	
	# Error log
	func error(owner: Object, msg: String):
		var context = owner.get_script().resource_path.get_file()
		print(formatted_message(context, msg, "E"))
		push_error(formatted_message(context, msg, "E"))
		if ESCProjectSettingsManager.get_setting(
			ESCProjectSettingsManager.TERMINATE_ON_ERRORS
		):
			assert(false)
			escoria.get_tree().quit()
```

StraToN replied:
Fixed already, thanks!

balloonpopper commented:
@StraToN It immediately breaks due to a bug. Please change "quitting" on line 339 in esc_object_manager.gd to "is_quitting" :
```
	if not escoria.is_quitting and escoria.inventory_manager.inventory_has(object.global_id):
```

StraToN commented:
@balloonpopper 

> It immediately breaks due to a bug. Please change "quitting" on line 339 in esc_object_manager.gd to "is_quitting" :

Oops. Sorry!

> Confirmed that you get a helpful error when the 2 objects share a global ID in room 12. The error message tells me that the child is the one that's registered rather than the parent, is this correct? I would've thought the parent would've been set up, then the child, and the child should complain it can't be set up as the parent is the one that's already configured. 

This is correct because our calls to `register_object()` happen in the `_ready()` function. Child's `_ready()` is called **before** its parent's. Check this link for more details: https://kidscancode.org/godot_recipes/basics/tree_ready_order/
If this is an issue, we may want to move these calls to `_enter_tree()` function instead.

BHSDuncan commented on addons/escoria-core/game/core-scripts/esc/esc_object_manager.gd:
```diff
@@ -483,14 +455,12 @@ func _room_exists(room_key: ESCRoomObjectsKey) -> bool:
 #
 # - object: The object to check for existence.
 # - room_key: The key representing the desired room in the object manager array.
-# **Returns** True iff object exists in the object manager entry specified by room_key.
```
Please put this back. "iff" isn't a typo; it's a common shorthand for "if and only if", which I've used on purpose.

StraToN replied:
oops! Alright then, sorry about that I had no idea of this :D

StraToN commented:
> If I start the game without a dialog plugin, it runs until I try to look at something and then says
"ERROR: ESC (2022-05-18T14:53:37) E esc_dialog_player.gd: No dialog manager called floating configured."
It would be nice if we could detect that there's no dialog manager at the start - I'm not sure if that's possible.
Everything works fine after reenabling the dialog plugin.

I added a verification step at the end of `escoria._ready()` function. We verify here that mandatory plugin(s) is(are) enabled. If not, end on error.

> If I disable the UI it gives the error
"ERROR: ESC (2022-05-18T15:02:36) E escoria.gd: Project setting 'escoria/ui/game_scene' is not set!"
This is correct as the game_scene is now empty. If I re-add the UI the game_scene automatically repopulates and everything works again. So things work fine, but shouldn't game_scene stay populated after you disable the UI plugin?

I chose to remove the setting value when the plugin is disabled because if we let it set, then the game will use the plugin even though it is disabled. You can try this by trying to re-add the setting manually while keeping the plugin disabled: the game runs fine and uses the disabled UI.

> I tried disabling/reenabling the plugins in lots of different orders and it all works fine. The only things I noticed :
** I can enable all 5 plugins and don't get any errors - should we error if there's more than one UI plugin?

When you enable a second UI plugin, there's an error that shows in the console:
```
 modules/gdscript/gdscript_functions.cpp:774 - Can't register user interface because user interface res://addons/escoria-ui-9verbs/game.tscn is registered.
```

You'll be happy: I managed to block UI plugin enabling if there is already one enabled.

> I disabled the Escoria plugin and left the other 2 enabled. Is it possible to make the error any more user friendly and ask the user to ensure the core plugin is enabled?

I managed to do this. Fixed (I think!)

balloonpopper commented:
Just refreshed to the latest commit. 
* On opening Escoria, this is in the console. I assume this isn't fixable as things are loading in?
![image](https://user-images.githubusercontent.com/5151242/169424246-f439a450-bbd5-45d2-b65d-1572a9d88cf5.png)
* The first thing I did was disable all 3 plugins from the top down. I then tried to enable all plugins from the top down. Core enabled successfully, but trying to enable either dialog or 9verb fails. The only way I can find to fix this is restore the project from Git.
![image](https://user-images.githubusercontent.com/5151242/169424508-f7a3afdd-47be-4dd5-a496-03e374ef38a6.png)
![image](https://user-images.githubusercontent.com/5151242/169424618-46d1c360-aab6-4318-a3e6-a099bf479453.png)
* Trying to enable a second UI plugin doesn't work, but it would be nice if you could get a popup in addition to the error in the console/debugger
* I understand this error when I disable the dialog plugin, but it doesn't make sense to me that I get the error when I reenable it
![image](https://user-images.githubusercontent.com/5151242/169424931-a43a6555-b94f-4c5f-bcc8-43f87390f5f2.png)
* Trying to run the game after disabling and reenabling the dialog/UI plugins fails.
![image](https://user-images.githubusercontent.com/5151242/169425024-0b8c97a4-d749-48d2-a61b-c40ef7a8fbd4.png)
![image](https://user-images.githubusercontent.com/5151242/169425053-43a0dde2-9226-4769-871e-079eb202488b.png)

StraToN commented:
@balloonpopper 

> On opening Escoria, this is in the console. I assume this isn't fixable as things are still loading in at this point?

Correct, nothing to be done there.

> The first thing I did was disable all 3 plugins from the top down. I then tried to enable all plugins from the top down. Core enabled successfully, but trying to enable either dialog or 9verb fails. The only way I can find to fix this is restore the project from Git.

Apparently no need to restore from Git. However, when this problem occurs, you can reload the project using the dedicated menu entry.
![image](https://user-images.githubusercontent.com/1289207/172999499-c2393899-c005-4f9d-ad9c-2381ce57275d.png)

> Trying to enable a second UI plugin doesn't work, but it would be nice if you could get a popup in addition to the error in the console/debugger

As per my tests, no this is not possible because we'd have to yield for the popup to be closed in order to disable the popup. In that case, the checkbox in the plugins list remains checked and you need to manually click the "Update" button to refresh that list (and have the checkbox unchecked).
If I don't yield for the popup to be closed, then the plugin is immediately disabled, even *before* the plugin can even show up.

StraToN commented:
> I understand this error when I disable the dialog plugin, but it doesn't make sense to me that I get the error when I reenable it

This is a bug in `ESCProjectSettingsManager.register_setting()` as this function was printing an error when the setting already exists. Will be fixed.

> Trying to run the game after disabling and reenabling the dialog/UI plugins fails.

I am not able to replicate. I tried to run both full game and specific room after disabling/reenabling both plugins, I had no issue and the game ran fine when both plugins are enabled. Can you be more specific?

BHSDuncan commented:
I couldn't reproduce the "trying to run the game after disabling and reenabling the dialog/UI plugins fails" issue, either. I tried unchecking and checking the dialog addon then running, and then unchecking and checking both the dialog and the 9verbs UI and then running. The game ran both times.

Like @StraToN was wondering, is there a more specific ordering of steps to reproduce the issue?

balloonpopper commented:
* I still had the issue on the latest commit with reenabling the plugin as mentioned in Discord, however after the initial switch from my branch to yours, and closing some scenes I had open, I haven't been able to reproduce it so we'll have to ignore that until it reoccurs. 
* I also saw a few error messages when I had Escoria open with my project, then closed Godot and checked out your branch.

Reproducing
* I had my animation helper files/scenes open in Godot (I was using my branch)
* Closed Godot
* Checked out your branch
* Make sure no files are showing in "git status" as different to your branch
* Open Escoria (It complained about one of my .gd files not being able to be opened as it no longer existed)
* I didn't touch the plugin settings
* Load then run scene 1 directly. This error appears at the start of the console output.
```
 Resource: 'res://addons/escoria-core/game/esc_autoload.gd' is already being loaded. Cyclic reference?
 res://addons/escoria-core/game/esc_logger.gd:60 - Parse Error: The identifier "escoria" isn't declared in the current scope.
 modules/gdscript/gdscript.cpp:577 - Method failed. Returning: ERR_PARSE_ERROR
 res://addons/escoria-core/game/esc_autoload.gd:36 - Parse Error: Couldn't fully preload the script, possible cyclic reference or compilation error. Use "load()" instead if a cyclic reference is intended.
 modules/gdscript/gdscript.cpp:577 - Method failed. Returning: ERR_PARSE_ERROR
```

* I stopped and restarted the scene, I saw lots of these errors (only in the console window, not in the debugger)
```
ERROR: Resource: 'res://addons/escoria-core/game/esc_autoload.gd' is already being loaded. Cyclic reference?
   at: (core/io/resource_loader.cpp:340)
SCRIPT ERROR: Parse Error: The identifier "escoria" isn't declared in the current scope.
          at: GDScript::reload (res://addons/escoria-core/game/esc_logger.gd:60)
ERROR: Method failed. Returning: ERR_PARSE_ERROR
   at: reload (modules/gdscript/gdscript.cpp:577)
SCRIPT ERROR: Parse Error: Couldn't fully preload the script, possible cyclic reference or compilation error. Use "load()" instead if a cyclic reference is intended.
          at: GDScript::reload (res://addons/escoria-core/game/esc_autoload.gd:36)
ERROR: Method failed. Returning: ERR_PARSE_ERROR
   at: reload (modules/gdscript/gdscript.cpp:577)
ERROR: Resource: 'res://addons/escoria-core/game/esc_autoload.gd' is already being loaded. Cyclic reference?
```

When I press ESC then quit, I still see these errors in the console
```
Closing logs peacefully.
ERROR: Condition "_first != nullptr" is true.
   at: ~List (./core/self_list.h:108)
ERROR: Condition "_first != nullptr" is true.
   at: ~List (./core/self_list.h:108)
ERROR: Condition "_first != nullptr" is true.
   at: ~List (./core/self_list.h:108)
WARNING: ObjectDB instances leaked at exit (run with --verbose for details).
     at: cleanup (core/object.cpp:2064)
ERROR: Resources still in use at exit (run with --verbose for details).
   at: clear (core/resource.cpp:417)
ERROR: There are still MemoryPool allocs in use at exit!
   at: (core/pool_vector.cpp:63)
```

balloonpopper commented:
1) First error
Testing starting point :
Opened room 7 in the Godot editor. Closed Godot. Checked out branch. Restarted Godot.
Errors in the console

```
PLUGIN ENTER TREE
PLUGIN _READY
SCRIPT ERROR: Invalid get index 'logger' (on base: 'Node (esc_autoload.gd)').
          at: ESCTerrain._ready (res://addons/escoria-core/game/core-scripts/esc_terrain.gd:73)
SCRIPT ERROR: Invalid get index 'room_manager' (on base: 'Node (esc_autoload.gd)').
          at: ESCRoom._ready (res://addons/escoria-core/game/core-scripts/esc_room.gd:69)
```

2) I like that I can't enable dialog/9 verb when core isn't enabled.
There's 5 of these errors in the console.
```
ERROR: Condition "!p_info.has("type")" is true.
   at: _add_property_info_bind (core/project_settings.cpp:939)
```
I get these errors whenever I disable simple dialog.

3) The following errors appear when I enable core
```
PLUGIN ENTER TREE
PLUGIN _READY
PLUGIN ENABLE_PLUGIN
ERROR: Resource: 'res://addons/escoria-core/game/esc_autoload.gd' is already being loaded. Cyclic reference?
   at: (core/io/resource_loader.cpp:340)
SCRIPT ERROR: Parse Error: The identifier "escoria" isn't declared in the current scope.
          at: GDScript::reload (res://addons/escoria-core/game/esc_logger.gd:60)
ERROR: Method failed. Returning: ERR_PARSE_ERROR
   at: reload (modules/gdscript/gdscript.cpp:577)
SCRIPT ERROR: Parse Error: Couldn't fully preload the script, possible cyclic reference or compilation error. Use "load()" instead if a cyclic reference is intended.
          at: GDScript::reload (res://addons/escoria-core/game/esc_autoload.gd:36)
ERROR: Method failed. Returning: ERR_PARSE_ERROR
   at: reload (modules/gdscript/gdscript.cpp:577)
ERROR: Script does not inherit a Node: res://addons/escoria-core/game/esc_autoload.gd.
   at: (editor/editor_autoload_settings.cpp:356)
ERROR: Condition "!info->node" is true. Continuing.
   at: update_autoload (editor/editor_autoload_settings.cpp:488)
DEFAULT DIALOG TYPE RESET !!!
```
No errors appear when I then enable dialog and 9 verb.

4) I now have all 3 plugins enabled and start room 7. In the console window I see the following errors 16 times
```
ERROR: Resource: 'res://addons/escoria-core/game/esc_autoload.gd' is already being loaded. Cyclic reference?
   at: (core/io/resource_loader.cpp:340)
SCRIPT ERROR: Parse Error: The identifier "escoria" isn't declared in the current scope.
          at: GDScript::reload (res://addons/escoria-core/game/esc_logger.gd:60)
ERROR: Method failed. Returning: ERR_PARSE_ERROR
   at: reload (modules/gdscript/gdscript.cpp:577)
SCRIPT ERROR: Parse Error: Couldn't fully preload the script, possible cyclic reference or compilation error. Use "load()" instead if a cyclic reference is intended.
          at: GDScript::reload (res://addons/escoria-core/game/esc_autoload.gd:36)
ERROR: Method failed. Returning: ERR_PARSE_ERROR
   at: reload (modules/gdscript/gdscript.cpp:577)
```

From further testing it might be caused if you disable all plugins, try to enable dialog/9 verb and then reenable everything. If you just disable them from the top, then reenable them from the top (without trying to enable dialog/9 verb inbetween) these errors don't appear.

5) I played the game to room 5, then quit using Esc and the quit menu. These errors are on the console (though I think this has been raised as a separate issue already)
```
ESC (2022-06-22T22:04:22) I esc_inputs_manager.gd: Item focused: r5_empty_sheet
ESC (2022-06-22T22:04:22) I esc_inputs_manager.gd: Item unfocused: r5_empty_sheet
Closing logs peacefully.
ERROR: Condition "_first != nullptr" is true.
   at: ~List (./core/self_list.h:108)
ERROR: Condition "_first != nullptr" is true.
   at: ~List (./core/self_list.h:108)
ERROR: Condition "_first != nullptr" is true.
   at: ~List (./core/self_list.h:108)
WARNING: ObjectDB instances leaked at exit (run with --verbose for details).
     at: cleanup (core/object.cpp:2064)
ERROR: Resources still in use at exit (run with --verbose for details).
   at: clear (core/resource.cpp:417)
ERROR: There are still MemoryPool allocs in use at exit!
   at: (core/pool_vector.cpp:63)
```

6) On your branch, I've played room 7, turned all the plugins off and on, and played room 7 again. These files are updated (which may or may not be expected)
```
        modified:   addons/escoria-core/game/main_scene.gd
        modified:   escoria.godot
        modified:   game/rooms/room07/room07.tscn
        modified:   project.godot
```

7) I ran the full game (F5, not F6) and quit from inside room 1. I noticed an error (Interrupt() function not implemented) in the debug output which may or may not be expected.
```
ESC (2022-06-22T22:12:48) I change_scene.gd: [change_scene] Changing scene to res://game/rooms/room01/room01.tscn (enable_automatic_transition = True).
ESC (2022-06-22T22:12:48) D esc_event_manager.gd: Interrupting running event newgame in channel _front...
ESC (2022-06-22T22:12:48) I esc_event.gd: Interrupting event newgame ([Reference:1778]).
ESC (2022-06-22T22:12:48) W change_scene.gd: [change_scene] Interrupt() function not implemented.
WARNING: ESC (2022-06-22T22:12:48) W change_scene.gd: [change_scene] Interrupt() function not implemented.
     at: call (modules/gdscript/gdscript_functions.cpp:788)
ESC (2022-06-22T22:12:48) D esc_compiler.gd: Parsing file res://game/rooms/room01/esc/right_exit.esc.
```

8) I restarted Godot, disabled all the plugins, then enabled core, dialog and simple_mouse. With simple_mouse the game crashes when I try to play a scene directly from game.gd
![image](https://user-images.githubusercontent.com/5151242/175027162-d9a58803-11bc-43b1-b385-7988b044d1c2.png)

Lastly) I'm not 100% sure how to create a game from scratch with these plugins, so I haven't tested that.

balloonpopper reviewed (approved):
A popup window to tell people to ignore any errors on the console and restart Godot to make them disappear would be handy, but otherwise it all looks good to me! Awesome work!!!

BHSDuncan commented on addons/escoria-core/game/core-scripts/esc/commands/anim.gd:
```diff
@@ -56,9 +55,7 @@ func run(command_params: Array) -> int:
-		]
+	escoria.logger.warn(
+		self,
+		"[%s] Interrupt() function not implemented." % get_command_name()
```
```suggestion
		"[%s] interrupt() method not implemented." % get_command_name()
```

This might be a bit picky, but it speaks to accuracy:

- "interrupt" should be lowercase since the method name itself is case sensitive and it is "interrupt()", not "Interrupt()"
- Technically "interrupt" is a method and not a function. (Actually the more I think about this, the less I care as it's probably a little too pedantic...Feel free to ignore this part of the suggestion.)

This of course would need to be changed across all commands that have this "not implemented" message.

BHSDuncan commented on addons/escoria-core/game/core-scripts/esc/commands/set_animations.gd:
```diff
@@ -27,19 +27,17 @@ func validate(arguments: Array):
-			]
+		escoria.logger.error(
+			self,
+			"[%s]: invalid animations. The animation resource %s was not found."
```
```suggestion
			"[%s]: invalid animation resource. The animation resource %s was not found."
```

BHSDuncan commented on addons/escoria-core/game/core-scripts/esc/esc_action_manager.gd:
```diff
@@ -344,16 +350,15 @@ func _activate(
+				escoria.logger.warn(
+					self,
+					"Invalid action on item" +
+					"Trying to run %s on object %s, " %
```
```suggestion
					"Trying to run action %s on object %s, " %
```

BHSDuncan commented on addons/escoria-core/game/core-scripts/esc/esc_compiler.gd:
```diff
@@ -246,34 +263,19 @@ func _compile(lines: Array, path: String = "") -> Array:
-					]
+				escoria.logger.error(
+					self,
+					"Command \"%s\" cannot be found under folder %s.\nPlease confirm setting \"%s\" is set to the folder where the ESC commands are found."
```
```suggestion
					"Command \"%s\" cannot be found under folder %s.\nPlease confirm setting \"%s\" is set to the folder where ESC commands are stored."
```

BHSDuncan commented on addons/escoria-core/game/core-scripts/esc/esc_globals_manager.gd:
```diff
@@ -117,6 +121,22 @@ func set_global_wildcard(pattern: String, value) -> void:
+#
+# #### Parameters
+#
+# * string: Text to log
```
```suggestion
# * string: Text in which to replace globals
```

...should probably also put in what the return value is.

BHSDuncan commented on addons/escoria-core/game/core-scripts/esc/esc_object_manager.gd:
```diff
@@ -430,13 +404,11 @@ func get_start_location() -> ESCLocation:
-		)
+	escoria.logger.warn(
+		self,
+		"Room has no ESCLocation node with 'is_start_location' enabled." +
```
```suggestion
		"Room has no ESCLocation node with 'is_start_location' enabled. " +
```

BHSDuncan commented on addons/escoria-core/game/core-scripts/save_data/esc_save_manager.gd:
```diff
@@ -444,10 +444,11 @@ func load_settings() -> Resource:
-			"Setting default settings."])
+		escoria.logger.warn(
+			self,
+			"Settings file %s doesn't exist. Setting default settings." 
```
```suggestion
			"Settings file %s doesn't exist. Using default settings." 
```

BHSDuncan commented on addons/escoria-core/game/esc_logger.gd:
```diff
@@ -0,0 +1,194 @@
+	signal perform_emergency_savegame
+	
+	# Log file format
+	const  LOG_FILE_FORMAT: String = "log_%s_%s.log"
```
```suggestion
	const LOG_FILE_FORMAT: String = "log_%s_%s.log"
```

BHSDuncan commented on addons/escoria-core/game/esc_logger.gd:
```diff
@@ -0,0 +1,194 @@
+class ESCLoggerBase:
```
One thing that we should get in the habit of doing is the ordering of code in a file, according to https://docs.godotengine.org/en/stable/tutorials/scripting/gdscript/gdscript_styleguide.html#code-order

I've been trying to update our code whenever I can to follow this, but I'd strongly recommend a good starting point being any new files we add in, e.g. this one, the new autoload file, etc.

BHSDuncan commented on addons/escoria-core/game/core-scripts/esc/esc_event_manager.gd:
```diff
@@ -353,25 +358,26 @@ func set_changing_scene(p_is_changing_scene: bool) -> void:
 	var event_flags = event.flags
 	if event_flags & ESCEvent.FLAG_NO_TT:
 		escoria.main.current_scene.game.tooltip_node.show()
+		pass
```
I think this can be removed.

BHSDuncan commented on addons/escoria-core/game/core-scripts/esc/esc_event_manager.gd:
```diff
@@ -353,25 +358,26 @@ func set_changing_scene(p_is_changing_scene: bool) -> void:
 
 	if event_flags & ESCEvent.FLAG_NO_UI:
 		escoria.main.current_scene.game.show_ui()
+		pass
```
I think this can be removed.

BHSDuncan reviewed (changes requested):
Nothing really major wrong here on first pass through code. In fact I found myself just getting more nitpicky as I kept reading, but all of that stuff can be addressed in the future.

Great work! Thank you for persevering!

StraToN commented:
@BHSDuncan Should be ok now :) Thanks for reviewing!

BHSDuncan commented on addons/escoria-core/game/core-scripts/esc/esc_event_manager.gd:
```diff
@@ -278,7 +278,7 @@ func queue_background_event(channel_name: String, event: ESCEvent) -> void:
 
 
-# Interrupt the events currently running and any that are pending.
+# interrupt() the events currently running and any that are pending.
```
Uh oh...I think your use of sed was a little overzealous. ;) Comments and error messages that now read "interrupt()" instead of "Interrupt" or "Interrupting", etc., should be put back. Sorry!

StraToN replied:
Fixed! Sorry about that.
